internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
 
-	func() {
-		c.subscribeFuncMu.Lock()
-		defer c.subscribeFuncMu.Unlock()
+	return nil
+}
 
-		if c.subscribeFunc != nil {
-			c.subscribeFunc(CompleterJobUpdated{Job: job, JobStats: stats})
-		}
-	}()
+func (c *InlineCompleter) Start(ctx context.Context) error { return nil }
 
-	return nil
+func (c *InlineCompleter) Stop() {
+	c.wg.Wait()
 }
 
-type AsyncJobCompleter struct {
+// A default concurrency of 100 seems to perform better a much smaller number
+// like 10, but it's quite dependent on environment (10 and 100 bench almost
+// identically on MBA when it's on battery power). This number should represent
+// our best known default for most use cases, but don't consider its choice to
+// be particularly well informed at this point.
+const asyncCompleterDefaultConcurrency = 100
+
+type AsyncCompleter struct {
 	baseservice.BaseService
+	withSubscribe
 
-	concurrency     uint32
-	exec            PartialExecutor
-	eg              *errgroup.Group
-	subscribeFunc   func(update CompleterJobUpdated)
-	subscribeFuncMu sync.Mutex
+	concurrency int
+	errGroup    *errgroup.Group
+	exec        PartialExecutor
+}
+
+func NewAsyncCompleter(archetype *baseservice.Archetype, exec PartialExecutor) *AsyncCompleter {
+	return newAsyncCompleterWithConcurrency(archetype, exec, asyncCompleterDefaultConcurrency)
 }
 
-func NewAsyncCompleter(archetype *baseservice.Archetype, exec PartialExecutor, concurrency uint32) *AsyncJobCompleter {
-	eg := &errgroup.Group{}
-	// TODO: int concurrency may feel more natural than uint32
-	eg.SetLimit(int(concurrency))
+func newAsyncCompleterWithConcurrency(archetype *baseservice.Archetype, exec PartialExecutor, concurrency int) *AsyncCompleter {
+	errGroup := &errgroup.Group{}
+	errGroup.SetLimit(concurrency)
 
-	return baseservice.Init(archetype, &AsyncJobCompleter{
+	return baseservice.Init(archetype, &AsyncCompleter{
 		exec:        exec,
 		concurrency: concurrency,
-		eg:          eg,
+		errGroup:    errGroup,
 	})
 }
 
-func (c *AsyncJobCompleter) JobSetStateIfRunning(stats *jobstats.JobStatistics, params *riverdriver.JobSetStateIfRunningParams) error {
-	return c.doOperation(stats, func(ctx context.Context) (*rivertype.JobRow, error) {
-		return c.exec.JobSetStateIfRunning(ctx, params)
-	})
-}
-
-func (c *AsyncJobCompleter) doOperation(stats *jobstats.JobStatistics, f func(ctx context.Context) (*rivertype.JobRow, error)) error {
-	c.eg.Go(func() error {
-		start := c.TimeNowUTC()
+func (c *AsyncCompleter) JobSetStateIfRunning(ctx context.Context, stats *jobstats.JobStatistics, params *riverdriver.JobSetStateIfRunningParams) error {
+	// Start clock outside of goroutine so that the time spent blocking waiting
+	// for an errgroup slot is accurately measured.
+	start := c.TimeNowUTC()
 
-		job, err := withRetries(&c.BaseService, f)
+	c.errGroup.Go(func() error {
+		job, err := withRetries(ctx, &c.BaseService, func(ctx context.Context) (*rivertype.JobRow, error) {
+			return c.exec.JobSetStateIfRunning(ctx, params)
+		})
```

Comment:
Hmm, is this an appropriate context to be passing down and respecting the deadline on? Isn't this `ctx` the job's execution context, which may be cancelled (including by a remote cancellation attempt)?

I feel like the only place it would make sense to be used is in the inline completer. The async one should probably be started with a Client-derived context that is in no way tied to the job's execution.